File: windows/tests/Elevate.Core.Tests/DiagnosticsReportTests.cs
using Elevate.Core.Support;
using FluentAssertions;

namespace Elevate.Core.Tests;

/// <summary>Port of the Swift <c>DiagnosticsReportTests</c> and <c>ErrorLogTests</c>.</summary>
public class DiagnosticsReportTests
{
    private static DiagnosticsInput MakeInput(
        IReadOnlyList<DiagnosticsError>? errors = null,
        IReadOnlyList<DiagnosticsAccount>? accounts = null,
        IReadOnlyList<DiagnosticsTenant>? tenants = null,
        IReadOnlyList<string>? profiles = null,
        string? hotKey = null) =>
        new("1.2.3", "45", "Unsigned", "Windows 11 26200", accounts ?? [], tenants ?? [], profiles ?? [], hotKey, errors ?? []);

    [Fact]
    public void HeaderLinesPresent()
    {
        var text = DiagnosticsReport.Render(MakeInput());
        text.Should().Contain("1.2.3").And.Contain("45").And.Contain("Unsigned").And.Contain("Windows 11 26200");
    }

    [Fact]
    public void AccountsSectionRendered()
    {
        var text = DiagnosticsReport.Render(MakeInput(accounts: [new("alice@contoso.com", "MSAL", 2)]));
        text.Should().Contain("alice@contoso.com").And.Contain("MSAL").And.Contain("2 tenant(s)");
    }

    [Fact]
    public void TenantsSectionRendered()
    {
        var text = DiagnosticsReport.Render(MakeInput(tenants: [new("Contoso", "tenant-id-1", "auto", ["azure", "groups"])]));
        text.Should().Contain("Contoso").And.Contain("tenant-id-1").And.Contain("auto").And.Contain("azure, groups");
    }

    [Fact]
    public void ProfilesAndHotKeyRendered()
    {
        var text = DiagnosticsReport.Render(MakeInput(profiles: ["Default", "Work"], hotKey: "Ctrl+Shift+E"));
        text.Should().Contain("Default").And.Contain("Work").And.Contain("Hot key: Ctrl+Shift+E");
    }

    [Fact]
    public void NoHotKeyRendersPlaceholder() =>
        DiagnosticsReport.Render(MakeInput()).Should().Contain("Hot key: None");

    [Fact]
    public void ErrorsRenderedWithIso8601UtcTimestamps()
    {
        var date = DateTimeOffset.FromUnixTimeSeconds(1_700_000_000);
        var text = DiagnosticsReport.Render(MakeInput(errors: [new(date, "boom")]));
        text.Should().Contain("[2023-11-14T22:13:20Z] boom");
    }

    [Fact]
    public void ErrorMessageWithSecretIsRenderedVerbatim() =>
        DiagnosticsReport.Render(MakeInput(errors: [new(DateTimeOffset.UtcNow, "auth failed token=abc123")]))
            .Should().Contain("token=abc123");

    [Fact]
    public void RendererNeverContainsUnpassedClientId() =>
        // The input type has no field for a client id / secret at all, so a client-id-looking
        // string that was never supplied cannot appear.
        DiagnosticsReport.Render(MakeInput()).Should().NotContain("9d3a7e2c-4b1f-4a6e-9c2d-5f8b1c0a7e3d");

    [Fact]
    public void MultipleErrorsRenderedInOrder()
    {
        var text = DiagnosticsReport.Render(MakeInput(errors:
        [
            new(DateTimeOffset.FromUnixTimeSeconds(100), "first error"),
            new(DateTimeOffset.FromUnixTimeSeconds(200), "second error"),
        ]));
        text.IndexOf("first error", StringComparison.Ordinal).Should().BePositive()
            .And.BeLessThan(text.IndexOf("second error", StringComparison.Ordinal));
    }

    [Fact]
    public void EmptySectionsSayNone()
    {
        var text = DiagnosticsReport.Render(MakeInput());
        text.Should().Contain("Accounts:\n  None").And.Contain("Tenants:\n  None")
            .And.Contain("Profiles:\n  None").And.Contain("Recent errors:\n  None");
    }

    [Fact]
    public void ErrorLogAppendsAndReturnsInOldestToNewestOrder()
    {
        var log = new ErrorLog();
        var d1 = DateTimeOffset.FromUnixTimeSeconds(1);
        var d2 = DateTimeOffset.FromUnixTimeSeconds(2);
        log.Append("first", d1);
        log.Append("second", d2);
        log.Entries.Select(e => e.Message).Should().Equal("first", "second");
        log.Entries.Select(e => e.Date).Should().Equal(d1, d2);
    }

    [Fact]
    public void ErrorLogCapsAtCapacityKeepingNewest()
    {
        var log = new ErrorLog(capacity: 3);
        for (var i = 0; i < 5; i++)
        {
            log.Append($"msg{i}", DateTimeOffset.FromUnixTimeSeconds(i));
        }

        log.Entries.Select(e => e.Message).Should().Equal("msg2", "msg3", "msg4");
    }

    [Fact]
    public void ErrorLogDefaultCapacityIs50()
    {
        var log = new ErrorLog();
        for (var i = 0; i < 60; i++)
        {
            log.Append($"msg{i}", DateTimeOffset.FromUnixTimeSeconds(i));
        }

        log.Entries.Should().HaveCount(50);
        log.Entries[0].Message.Should().Be("msg10");
        log.Entries[^1].Message.Should().Be("msg59");
    }

    [Fact]
    public void ErrorLogContentEquality()
    {
        var a = new ErrorLog(capacity: 5);
        var b = new ErrorLog(capacity: 5);
        var d = DateTimeOffset.FromUnixTimeSeconds(10);
        a.Append("x", d);
        b.Append("x", d);
        a.ContentEquals(b).Should().BeTrue();
        b.Append("y", d);
        a.ContentEquals(b).Should().BeFalse();
    }
}

File: macos/Tests/ElevateCoreTests/DiagnosticsReportTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct DiagnosticsReportTests {
    private func makeInput(
        errors: [DiagnosticsError] = [],
        accounts: [DiagnosticsAccount] = [],
        tenants: [DiagnosticsTenant] = [],
        profiles: [String] = [],
        hotKey: String? = nil
    ) -> DiagnosticsInput {
        DiagnosticsInput(
            appVersion: "1.2.3",
            build: "45",
            signing: "Developer ID",
            os: "macOS 15.1",
            accounts: accounts,
            tenants: tenants,
            profiles: profiles,
            hotKey: hotKey,
            errors: errors
        )
    }

    @Test func headerLinesPresent() {
        let text = DiagnosticsReport.render(makeInput())
        #expect(text.contains("1.2.3"))
        #expect(text.contains("45"))
        #expect(text.contains("Developer ID"))
        #expect(text.contains("macOS 15.1"))
    }

    @Test func accountsSectionRendered() {
        let input = makeInput(accounts: [
            DiagnosticsAccount(upn: "alice@contoso.com", method: "MSAL", tenantCount: 2)
        ])
        let text = DiagnosticsReport.render(input)
        #expect(text.contains("alice@contoso.com"))
        #expect(text.contains("MSAL"))
        #expect(text.contains("2"))
    }

    @Test func tenantsSectionRendered() {
        let input = makeInput(tenants: [
            DiagnosticsTenant(name: "Contoso", id: "tenant-id-1", mode: "auto", flags: ["azure", "groups"])
        ])
        let text = DiagnosticsReport.render(input)
        #expect(text.contains("Contoso"))
        #expect(text.contains("tenant-id-1"))
        #expect(text.contains("auto"))
        #expect(text.contains("azure"))
        #expect(text.contains("groups"))
    }

    @Test func profilesAndHotKeyRendered() {
        let input = makeInput(profiles: ["Default", "Work"], hotKey: "cmd+shift+e")
        let text = DiagnosticsReport.render(input)
        #expect(text.contains("Default"))
        #expect(text.contains("Work"))
        #expect(text.contains("cmd+shift+e"))
    }

    @Test func noHotKeyRendersPlaceholder() {
        let input = makeInput(hotKey: nil)
        let text = DiagnosticsReport.render(input)
        #expect(text.contains("None") || text.contains("none") || text.contains("—"))
    }

    @Test func errorsRenderedWithISO8601UTCTimestamps() {
        let date = Date(timeIntervalSince1970: 1_700_000_000)
        let input = makeInput(errors: [DiagnosticsError(date: date, message: "boom")])
        let text = DiagnosticsReport.render(input)
        #expect(text.contains("boom"))
        // ISO-8601 UTC representation of the fixed timestamp above.
        let formatter = ISO8601DateFormatter()
        formatter.timeZone = TimeZone(identifier: "UTC")
        let expected = formatter.string(from: date)
        #expect(text.contains(expected))
    }

    @Test func errorMessageWithSecretIsRenderedVerbatim() {
        let input = makeInput(errors: [
            DiagnosticsError(date: .now, message: "auth failed token=abc123")
        ])
        let text = DiagnosticsReport.render(input)
        #expect(text.contains("token=abc123"))
    }

    @Test func rendererNeverContainsUnpassedClientId() {
        // The input type has no field for a client id / secret at all, so a
        // client-id-looking string that was never supplied cannot appear.
        let clientIdLooking = "9d3a7e2c-4b1f-4a6e-9c2d-5f8b1c0a7e3d"
        let text = DiagnosticsReport.render(makeInput())
        #expect(!text.contains(clientIdLooking))
    }

    @Test func multipleErrorsRenderedInOrder() {
        let d1 = Date(timeIntervalSince1970: 100)
        let d2 = Date(timeIntervalSince1970: 200)
        let input = makeInput(errors: [
            DiagnosticsError(date: d1, message: "first error"),
            DiagnosticsError(date: d2, message: "second error")
        ])
        let text = DiagnosticsReport.render(input)
        let firstRange = text.range(of: "first error")
        let secondRange = text.range(of: "second error")
        #expect(firstRange != nil && secondRange != nil)
        if let f = firstRange, let s = secondRange {
            #expect(f.lowerBound < s.lowerBound)
        }
    }
}
